core/store/src/trie/iterator.rs

```diff
@@ -442,27 +442,138 @@ mod tests {
+    /// included and the last one is excluded for boundary is leaf node
+    #[test]
+    fn test_visit_interval_boundary_at_leaf_node() {
+        init_test_logger();
+        let trie_changes = vec![
+            (b"aa".to_vec(), Some(vec![0])),
+            (b"abb".to_vec(), Some(vec![1])),
+            (b"abc".to_vec(), Some(vec![1])),
+            (b"b".to_vec(), Some(vec![2])),
+            (b"baab".to_vec(), Some(vec![3])),
+            (b"baac".to_vec(), Some(vec![4])),
+            (b"bbb".to_vec(), Some(vec![5])),
+        ];
+        let tries = create_tries();
+        let state_root =
+            test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
+        let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
+
+        // begin and end both at leaf nodes
+        verify_items_in_interval(
+            b"abb",
+            b"baac",
+            &trie,
+            vec![b"abb".to_vec(), b"abc".to_vec(), b"b".to_vec(), b"baab".to_vec()],
+        );
+    }
+
+    /// for boundary at extension node
+    /// Checks that for visiting interval of trie nodes first state key is
+    /// included and the last one is excluded
+    #[test]
+    fn test_visit_interval_boundary_at_extension_node() {
+        init_test_logger();
+        let trie_changes = vec![
+            (b"aa".to_vec(), Some(vec![0])),
+            (b"abb".to_vec(), Some(vec![1])),
+            (b"baab".to_vec(), Some(vec![3])),
+            (b"bab".to_vec(), Some(vec![4])),
+            (b"bbb".to_vec(), Some(vec![5])),
+        ];
+        let tries = create_tries();
+        let state_root =
+            test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
+        let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
+
+        // begin at an extension node, end at a leaf node
+        verify_items_in_interval(b"a", b"baab", &trie, vec![b"aa".to_vec(), b"abb".to_vec()]);
+    }
+
+    /// for boundary at branch nodes:
+    /// Checks that for visiting interval of trie nodes first state key is
+    /// included and the last one is excluded
+    #[test]
+    fn test_visit_interval_boundary_at_branch_node() {
+        init_test_logger();
+        let trie_changes = vec![
+            (b"aa".to_vec(), Some(vec![0])),
+            (b"abb".to_vec(), Some(vec![1])),
+            (b"abc".to_vec(), Some(vec![1])),
+            (b"b".to_vec(), Some(vec![2])),
+            (b"baab".to_vec(), Some(vec![3])),
+            (b"baac".to_vec(), Some(vec![4])),
+            (b"bbb".to_vec(), Some(vec![5])),
+        ];
+        let tries = create_tries();
+        let state_root =
+            test_populate_trie(&tries, &Trie::EMPTY_ROOT, ShardUId::single_shard(), trie_changes);
+        let trie = tries.get_trie_for_shard(ShardUId::single_shard(), state_root);
+
+        // begin at a branch node, end at a leaf node
+        verify_items_in_interval(
+            b"b",
```

Comment:
Does `b"b"` correspond to a branch node?
In the previous test you expect `b"a"` to correspond to an extension node.

`b` is `62` in Hex, has two nibbles: `6` and `2`.
My understanding is that `b` is a extension node for the nibble `6`, and its child node is a branch for nibbles `1` or `2`, leading to `ba` and `bb`.

Maybe replacing all letters `b` with letters `p` will make it a branch, as `p` is `70` in Hex.